cooked-validators/src/Cooked/Tx/Constraints/Type.hs

```diff
@@ -48,281 +48,230 @@ import Type.Reflection
-            pkOutputHashValue = value
-          }
-      )
-pkOutputWithDatumHashFromGeneric _ = Nothing
-
--- ** Outputs with datum belonging to scripts
-
--- I can't define this as a tuple, because otherwise I'll get an error "Illegal
--- type synonym family application in instance" concerning the 'Pl.DatumType' in
--- the 'IsOutput' instance declaration.
-data ScriptOutputWithDatum a where
-  ScriptOutputWithDatum ::
-    (Show (Pl.DatumType a), Pl.ToData (Pl.DatumType a)) =>
-    { scriptOutputDatumValidator :: Pl.TypedValidator a,
-      scriptOutputDatumCredential :: Maybe Pl.StakingCredential,
-      scriptOutputDatumDatum :: Pl.DatumType a,
-      scriptOutputDatumValue :: Pl.Value
-    } ->
-    ScriptOutputWithDatum a
+instance ToOutputDatum () where
+  toOutputDatum = const Pl.NoOutputDatum
 
-deriving instance (Show (ScriptOutputWithDatum a))
+instance ToOutputDatum Pl.DatumHash where
+  toOutputDatum = Pl.OutputDatumHash
 
-makeLensesFor
-  [ ("scriptOutputDatumValidator", "scriptOutputDatumValidatorL"),
-    ("scriptOutputDatumCredential", "scriptOutputDatumCredentialL"),
-    ("scriptOutputDatumDatum", "scriptOutputDatumDatumL"),
-    ("scriptOutputDatumValue", "scriptOutputDatumValueL")
-  ]
-  ''ScriptOutputWithDatum
-
-instance IsOutput (ScriptOutputWithDatum a) where
-  type OwnerType (ScriptOutputWithDatum a) = Pl.TypedValidator a
-  type DatumType (ScriptOutputWithDatum a) = Pl.DatumType a
-  sOutOwnerL = scriptOutputDatumValidatorL
-  sOutDatumL = scriptOutputDatumDatumL
-  sOutValueL = scriptOutputDatumValueL
-  sOutAddressL =
-    lens
-      (\output -> Pl.Address (Pl.ScriptCredential . Pl.validatorHash $ output ^. scriptOutputDatumValidatorL) (output ^. scriptOutputDatumCredentialL))
-      (\(ScriptOutputWithDatum _ _ datum value) (validator, mStCred) -> ScriptOutputWithDatum validator mStCred datum value)
-  sOutOutputDatumL =
-    lens
-      (\output -> Pl.OutputDatum (Pl.Datum . Pl.toBuiltinData $ (output ^. scriptOutputDatumDatumL)))
-      (flip (scriptOutputDatumDatumL .~))
-
--- TODO
-scriptOutputWithDatumFromGeneric :: GenericOutput -> Maybe (ScriptOutputWithDatum a)
-scriptOutputWithDatumFromGeneric = undefined
-
--- ** outputs with datum hash belonging to scripts
-
-data ScriptOutputWithDatumHash a where
-  ScriptOutputWithDatumHash ::
-    Show (Pl.DatumType a) =>
-    Pl.TypedValidator a ->
-    Maybe Pl.StakingCredential ->
-    Pl.DatumType a ->
-    Pl.Value ->
-    ScriptOutputWithDatumHash a
-
-deriving instance (Show (ScriptOutputWithDatumHash a))
-
-instance IsOutput (ScriptOutputWithDatumHash a) where
-  type OwnerType (ScriptOutputWithDatumHash a) = Pl.TypedValidator a
-  type DatumType (ScriptOutputWithDatumHash a) = Pl.DatumType a
-  sOutOwnerL = undefined
-  sOutDatumL = undefined
-  sOutValueL = undefined
-  sOutAddressL = undefined
-
-  -- TODO Remember that an 'OutputDatumHash ...' should be targeted by the lens
-  sOutOutputDatumL = undefined
+data ConcreteOutput ownerType datumType where
```

Comment:
A very small nitpick, but I'd probably remove the `Type` suffix in such contexts. It follows from the context that it's indeed a type, so it'll give shorter identifiers and less things to type.

But that's largely a stylistic preference, so please feel free to not bother with this!